scipy/special/tests/test_support_alternative_backends.py

```diff
@@ -21,83 +21,167 @@
     else:
-        positive_only = False
+        positive_only = nfo.positive_only
+    if isinstance(positive_only, bool):
+        positive_only = [positive_only]*nfo.n_args
+
+    if f_name in {'betaincinv'} and is_cupy(xp):
+        pytest.xfail("CuPy uses different convention for out of domain input.")
+
+    if (
+            f_name == "sinc" and "float32" in dtypes
+            and version.parse(np.__version__) < version.parse("2")
+    ):
+        pytest.xfail("https://github.com/numpy/numpy/issues/11204")
 
     if not any('int' in dtype for dtype in dtypes):
-        return positive_only, dtypes
+        return positive_only
 
     # Integer-specific issues from this point onwards
 
+    if f_name in {'gamma', 'gammasgn'} and is_cupy(xp):
+        # CuPy has not yet updated gamma pole behavior to match
+        # https://github.com/scipy/scipy/pull/21827.
+        positive_only = [True]
+
+    if f_name in {'poch'} and is_jax(xp):
+        # Jax uses a different convention at gamma poles.
+        positive_only = [True, True]
+
+    if f_name == 'multigammaln':
+        pytest.skip("multigammaln raises for out of domain inputs.")
+
     if ((is_torch(xp) and f_name in {'gammainc', 'gammaincc'})
         or (is_cupy(xp) and f_name in {'stdtr', 'i0e', 'i1e'})
-        or (is_jax(xp) and f_name in {'stdtr', 'ndtr', 'ndtri', 'log_ndtr'})
+        or (is_jax(xp) and f_name in {'stdtr', 'ndtr', 'ndtri', 'log_ndtr', 'hyp1f1',
+                                      'hyp2f1', 'spence', 'kl_div'})
     ):
         pytest.skip(f"`{f_name}` does not support integer types")
 
     # int/float mismatched args support is sketchy
     if (any('float' in dtype for dtype in dtypes)
-        and ((is_torch(xp) and f_name in ('rel_entr', 'xlogy'))
-             or (is_jax(xp) and f_name in ('gammainc', 'gammaincc',
-                                           'rel_entr', 'xlogy')))
+        and ((is_torch(xp) and f_name in ('rel_entr', 'xlogy', 'polygamma',
+                                          'zeta', 'xlog1py'))
+             or (is_jax(xp) and f_name in ('gammainc', 'gammaincc', 'expn',
+                                           'rel_entr', 'xlogy', 'betaln',
+                                           'polygamma', 'zeta', 'poch',
+                                           'xlog1py')))
     ):
         pytest.xfail("dtypes do not match")
 
-    dtypes_np_ref = dtypes
-    if (is_torch(xp) and xp_default_dtype(xp) == xp.float32
-        and f_name not in {'betainc', 'betaincc', 'stdtr', 'stdtrit'}
-    ):
-        # On PyTorch with float32 default dtype, sometimes ints are promoted
-        # to float32, and sometimes to float64.
-        # When they are promoted to float32, explicitly convert the reference
-        # numpy arrays to float32 to prevent them from being automatically promoted
-        # to float64 instead.
-        dtypes_np_ref = ['float32' if 'int' in dtype else dtype for dtype in dtypes]
```

Comment:
This handled the different promotion rules by modifying the inputs, but doesn't work for instance when there are integer only arguments. I changed it to modify the output to the correct type instead, directly at the point where the reference output is computed.